Fix docker swarm mode

mathcantin opened the pull request:
Fixes #216 

The fix is instead of using service.update(image=tag) we need to use service.update(image=f"{tag}@sha256:{latest_image_sha256}").

With a little bit of refractor to remove duplicate code.

mathcantin commented:
On the issues, there are this comment
> as another way to look at it he is saying it tries to update something that doesnt need updating, which is odd since its checking the sha when it DOES exist initially. Maybe the root can be found there

Still a good idea, but that not cover by this PR.

DirtyCajunRice commented:
have you ran this code? i like the baseobject but a lot of stuff just wont work. e.g. digest or something doesnt return digest or something. it returns true or false

mathcantin commented:
Yes, I try with and without swarm mode and that works. The method _get_digest, return the sha256 of the image, in all cases (dryrun on or off with auth or not). I don't know where you see a method that return a boolean.

mathcantin commented:
I understand, the code:
```
        digest = image.attrs.get("Descriptor", {}).get("digest")
        digest = digest or image.attrs.get("RepoDigests")[0].split('@')[1]
        digest = digest or image.id

        return self._remove_sha_prefix(digest)
```

is exactly the same thing then that

``` 
    digest = image.attrs.get("Descriptor", {}).get("digest")
    if not digest:
            digest = image.attrs.get("RepoDigests")[0].split('@')[1]
    if not digest
        digest = image.id

        return self._remove_sha_prefix(digest)
```

Because the or keyword isn't a comparator and don't return True or False. I can change it, if you prefer the second solution.

DirtyCajunRice commented:
Yeah i spoke before i ran the code myself. its completely fine. Im reading over the change right now. Give me a bit is a large change.

DirtyCajunRice commented on pyouroboros/dockerclient.py:
```diff
@@ -381,38 +383,24 @@ def monitor_filter(self):
-                raise ConnectionError
+        return self._pull(tag)
+
+    def _remove_sha_prefix(self, digest):
```
This is only called in one place, so why is it abstracted to its own private function?

mathcantin replied:
I use it two times. That improve readability of code.
